File: src/main/javatokotlin/user/Repository.java
package main.javatokotlin.user;

import java.util.ArrayList;
import java.util.List;

public class Repository {

    private static Repository INSTANCE = null;

    public List<User> users;

    // keeping the constructor private to enforce the usage of getInstance
    private Repository() {
        User user1 = new User("Chinedu", "Ihedioha");
        User user2 = new User("Uche", "");
        User user3 = new User("Olutoba", null);

        users = new ArrayList<>();
        users.add(user1);
        users.add(user2);
        users.add(user3);
    }

    public static Repository getInstance() {
        if (INSTANCE == null) {
            synchronized (Repository.class) {
                if (INSTANCE == null) {
                    INSTANCE = new Repository();
                }
            }
        }
        return INSTANCE;
    }

    public List<User> getUsers() {
        return users;
    }

    public List<String> getFormattedUserName() {
        List<String> userNames = new ArrayList<>(users.size());
        for (User user : users) {
            String name;
            if (user.getLastName() != null) {
                if (user.getFirstName() != null) {
                    name = user.getFirstName() + " " + user.getLastName();
                } else {
                    name = user.getLastName();
                }
            } else if (user.getFirstName() != null) {
                name = user.getFirstName();
            } else {
                name = "unknown";
            }
            userNames.add(name);
        }
        return userNames;
    }

}

File: src/main/javatokotlin/user/Repository.kt
package main.javatokotlin.user

object Repository {

    private val _users =
        mutableListOf(User("Chinedu", "Ihedioha"), User("Uche", ""), User("Olutoba", null))

    val users: List<User>
        get() = _users

    val formattedUserNames: List<String?>
        get() = _users.map { user -> user.userFormattedName }
}

//    init {
//        val user1 = User("Chinedu", "Ihedioha")
//        val user2 = User("Uche", "")
//        val user3 = User("Olutoba", null)
//
//        _users.apply {
//            add(user1)
//            add(user2)
//            add(user3)
//        }
//    }

// extension function
fun User.getFormattedName(): String {
    return if (lastName != null) {
        if (firstName != null) {
            "$firstName $lastName"
        } else {
            lastName ?: "Unknown"
        }
    } else {
        firstName ?: "unknown"
    }
}

// extension property
val User.userFormattedName: String
    get() {
        return if (lastName != null) {
            if (firstName != null) {
                "$firstName $lastName"
            } else {
                lastName ?: "Unknown"
            }
        } else {
            firstName ?: "Unknown"
        }
    }

val user = User("Olutoba ", "Onikoyi")
val name = user.getFormattedName()
val formattedName = user.userFormattedName